bootloader/piboot.go

```diff
@@ -0,0 +1,395 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package bootloader
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/bootloader/ubootenv"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+)
+
+// sanity - piboot implements the required interfaces
+var (
+	_ Bootloader                             = (*piboot)(nil)
+	_ ExtractedRecoveryKernelImageBootloader = (*piboot)(nil)
+)
+
+const (
+	pibootCfgFilename = "piboot.conf"
+	pibootPartFolder  = "/piboot/ubuntu/"
+	// TODO Use boot.InitramfsUbuntuSeedDir
+	// Need to solve circular dependency
+	runMntDir     = "/run/mnt/"
+	ubuntuSeedDir = runMntDir + "ubuntu-seed/"
```

Comment:
this is not great here indeed

Reply:
I think that the easiest way to solve the circular dependency would be to create an `initramfsconst` package in `boot/initramfs/` folder, I can add that to the PR if that works for you.

Reply:
The issue is not just the circular import. Is the conceptual problem that this bootloaders needs to know about something that no bootloader needs to know or mostly needs to know so far. So we need to decide whether we are ok or can do something else with the layer violation instead.

Reply:
The main reason for needing the path to ubuntu-seed is that kernel assets must be extracted there, and also bootloader configuration must be created there, not in ubuntu-boot, and the `rootdir` provided creating a new bootloader does not always point there. So, if using directly the variable is undesired we would need a way to get the path from the bootloader in some other way.

Reply:
Do I understand correctly, that if ubuntu-boot were the first partition in the disk, then we wouldn't need to touch ubuntu-seed here?

If so, the next question would be: is there a reason to keep ubuntu-seed as the first partition, or is it something that we could change? Maybe let devices specify the order of partitions (in case we have other devices which require ubuntu-seed to be the first partition)?

Reply:
I think we just want to leave a TODO here for now, the issue is really that while we must never need to touch the run mode bootloader state when dealing with the recovery bootloader, the reverse might not be true in general so the correct thing is really to pass more information that just a rootdir when we create the bootloader but that is largeish, probably best done later.